crates/lair_keystore_api/src/config.rs

```diff
@@ -0,0 +1,254 @@
+
+            // lock the context key
+            let context_key = SecretDataSized::encrypt(
+                ctx_secret.to_read_sized(),
+                context_key.to_read_sized(),
+            )
+            .await?;
+
+            // lock the signature seed
+            let sign_seed = SecretDataSized::encrypt(
+                sig_secret.to_read_sized(),
+                sign_seed.to_read_sized(),
+            )
+            .await?;
+
+            // get the signature public key bytes for encoding in the url
+            let sign_pk: BinDataSized<32> =
+                sign_pk.try_unwrap_sized().unwrap().into();
+
+            // on windows, we default to using "named pipes"
+            #[cfg(windows)]
+            let connection_url = {
+                let id = nanoid::nanoid!();
+                url::Url::parse(&format!(
+                    "named-pipe:\\\\.\\pipe\\{}?k={}",
+                    id, sign_pk
+                ))
+                .unwrap()
+            };
+
+            // on not-windows, we default to using unix domain sockets
+            #[cfg(not(windows))]
+            let connection_url = {
+                let mut con_path = root_path.clone();
+                con_path.push("socket");
+                url::Url::parse(&format!(
+                    "unix://{}?k={}",
+                    con_path.to_str().unwrap(),
+                    sign_pk
+                ))
+                .unwrap()
+            };
+
+            // put together the full server config struct
+            let config = LairServerConfigInner {
+                connection_url,
+                pid_file,
+                store_file,
+                runtime_secrets_salt: salt.try_unwrap_sized().unwrap().into(),
+                runtime_secrets_mem_limit: mem_limit,
+                runtime_secrets_ops_limit: ops_limit,
+                runtime_secrets_context_key: context_key,
+                runtime_secrets_sign_seed: sign_seed,
+            };
+
+            Ok(config)
+        }
+    }
+
+    /// Get the connection "scheme". i.e. "unix", "named-pipe", or "tcp".
+    pub fn get_connection_scheme(&self) -> &str {
+        self.connection_url.scheme()
+    }
+
+    /// Get the connection "path". This could have different meanings
+    /// depending on if we are a unix domain socket or named pipe, etc.
+    pub fn get_connection_path(&self) -> &str {
+        self.connection_url.path()
+    }
+
+    /// Get the server pub key BinDataSized<32> bytes from the connectionUrl
+    pub fn get_server_pub_key(&self) -> LairResult<BinDataSized<32>> {
+        get_server_pub_key_from_connection_url(&self.connection_url)
+    }
+}
+
+/// extract a server_pub_key from a connection_url
+pub fn get_server_pub_key_from_connection_url(
+    url: &url::Url,
+) -> LairResult<BinDataSized<32>> {
+    for (k, v) in url.query_pairs() {
+        if k == "k" {
+            let tmp =
+                base64::decode_config(v.as_bytes(), base64::URL_SAFE_NO_PAD)
```

Comment:
worth aliasing the base64 flavour as our own constant so we can't get it wrong anywhere?

Reply:
added alias functions in encoding_types plus FromStr impl on the BinData* types so this can just become a `parse()` call.